presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java

```diff
@@ -494,6 +498,28 @@ public Optional<Tracer> getTracer()
         return tracer;
     }
 
+    @Override
+    public void setAuthorizedIdentity(Identity identity, Optional<String> selectedUser, Optional<String> reasonForSelect)
+    {
+        authorizedIdentity = Optional.of(new Identity(
+                identity.getUser(),
+                identity.getPrincipal(),
+                identity.getRoles(),
+                identity.getExtraCredentials(),
+                identity.getExtraAuthenticators(),
+                selectedUser,
+                reasonForSelect));
+    }
+
+    @Override
+    public Session appendAuthorizedIdentity(Session session)
+    {
+        if (!authorizedIdentity.isPresent()) {
+            return session;
+        }
+        return Session.builder(session).setIdentity(authorizedIdentity.get()).build();
```

Comment:
This looks hacky in 2 ways.
1.  This method overwrites session and creates a new object. Some others could be holding on to the old session. Session is not really meant to be a mutable object.
2. the method name append sounds like it is mutating the current session object, but it is producing a new session object.

Reply:
I agree, Lin, if you move the security checks before creating the session object, this should help in clean up.

Reply:
Yeah, immutability is good point. I will modify accordingly.